### DISCUSSION TITLE
(NextJS)NextAuth tutorial - Image URL uploads as 'undefined'

### DISCUSSION THREAD

**Discussion by blueMug-Web** [supabase/supabase#10047] [Questions] [answered]
Hey everyone,

I have been working on the tutorial from NextAuth, the 'SupaVacation' project where people can create an account and save vacation spots, upload images, etc...

I have the site working, and the auth part of it is working as intended. However, the image-uploading part is not working for me yet. Image URLs that get created when the image is uploaded are ending up as 'undefined' instead of the name or unique id of the image.

Every 'vacation spot' that gets created has an id, image, title, description, price, guests, etc...

In the pages/api folder, I have a file 'image-upload.js', and there is a line of code that creates the url for the image when the image is uploaded. The image itself gets uploaded, I can see it in the bucket in Supabase, but when I check the Table, the url that is created for the image ends up looking like this 'https://<SUPABASE_URL>/storage/v1/object/public/undefined'

Instead of 'undefined', it should be a unique ID. When checking the bucket, the name of the image that is uploaded has what looks like a unique id, and so the url for the image file should look something like this: 'https://<SUPABASE_URL>/storage/v1/object/public/xK3qaRUBZRwvI_qStygXT.jpeg'

When I run the dev server in vsCode, the error I receive is this:
'upstream image response failed for https://<SUPABASE_URL>/storage/v1/object/public/undefined 400'

here is what the code looks like in my 'image-upload' file:

```
import { supabase } from "@/lib/supabase";
import { nanoid } from "nanoid";
import { decode } from "base64-arraybuffer";

export const config = {
	api: {
		bodyParser: {
			sizeLimit: "10mb",
		},
	},
};

export default async function handler(req, res) {
	// Upload image to Supabase
	if (req.method === "POST") {
		let { image } = req.body;

		if (!image) {
			return res.status(500).json({ message: "No image provided" });
		}

		try {
			const contentType = image.match(/data:(.*);base64/)?.[1];
			const base64FileData = image.split("base64,")?.[1];

			if (!contentType || !base64FileData) {
				return res.status(500).json({ message: "Image data not valid" });
			}

			// Upload image
			const fileName = nanoid();
			const ext = contentType.split("/")[1];
			const path = `${fileName}.${ext}`;

			const { data, error: uploadError } = await supabase.storage
				.from(process.env.SUPABASE_BUCKET)
				.upload(path, decode(base64FileData), {
					contentType,
					upsert: true,
				});

			if (uploadError) {
				console.log(uploadError);
				throw new Error("Unable to upload image to storage");
			}

			// ***NOTE: This is tutorial code, they needed to use replace, we do not. ***
			// Construct public URL
			// const url = `${process.env.SUPABASE_URL.replace(
			// 	".co",
			// 	".in"
			// )}/storage/v1/object/public/${data.Key}`;

			const url = `${process.env.SUPABASE_URL}/storage/v1/object/public/${data.Key}`;
			// *** THIS LINE ABOVE is not creating a proper URL, right now data.Key is being logged in supabase storage(bucket) as ../undefined ***

			return res.status(200).json({ url });
		} catch (e) {
			res.status(500).json({ message: "Something went wrong" });
		}
	}
	// HTTP method not supported!
	else {
		res.setHeader("Allow", ["POST"]);
		res
			.status(405)
			.json({ message: `HTTP method ${req.method} is not supported.` });
	}
}
```

**Comment by GaryAustin1** [comment, marked as answer]
If you are using supabase-js v2 (2.0.x) then storage return has changed and returns path instead of key

_The upload and update function returns the path of the object uploaded as the path parameter. Previously the returned value had the bucket name prepended to the path which made it harder to pass the value on to other storage-js methods since all methods take the bucket name and path separately. We also chose to call the returned value path instead of Key ([PR](https://github.com/supabase/storage-js/pull/75))_

**Reply by blueMug-Web** [reply]
Thank you! Issue resolved.
